sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodShouldBeNameAccordingToSynchronicity.cs

```diff
@@ -0,0 +1,90 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class MethodShouldBeNameAccordingToSynchronicity : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4261";
+        private const string MessageFormat = "{0}";
+        private const string AddAsyncSuffixMessage = "Add the 'Async' suffix to the name of this method.";
+        private const string RemoveAsyncSuffixMessage = "Remove the 'Async' suffix to the name of this method.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static ISet<KnownType> TaskTypes = new HashSet<KnownType>
+        {
+            KnownType.System_Threading_Tasks_Task,
+            KnownType.System_Threading_Tasks_Task_T
+        };
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var methodDeclaration = (MethodDeclarationSyntax)c.Node;
+                    if (methodDeclaration.Identifier.IsMissing)
+                    {
+                        return;
+                    }
+
+                    var methodSymbol = c.SemanticModel.GetDeclaredSymbol(methodDeclaration);
+                    if (methodSymbol == null ||
+                        methodSymbol.GetInterfaceMember() != null ||
+                        methodSymbol.GetOverriddenMember() != null)
+                    {
+                        return;
+                    }
+
+                    var isTaskReturnType = methodSymbol.ReturnType.DerivesFromAny(TaskTypes);
+                    var hasAsyncSuffix = "async".Equals(methodDeclaration.Identifier.ValueText.SplitCamelCaseToWords().LastOrDefault());

```

Comment:
Why not `async ==`?

Reply:
Good question! I will use `==`.

Reply:
@valhristov sadly the `SplitCamelCaseToWords` is changing the case (in the wrong way `ToLowerInvariant`)

Reply:
If it was case-insensitive, we may get (granted, unlikely) FPs on things like `GigaSync`.